spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TemporalMaxPooling.scala

```diff
@@ -0,0 +1,290 @@
+
+import com.intel.analytics.bigdl.nn.abstractnn.TensorModule
+import com.intel.analytics.bigdl.tensor.{DoubleType, FloatType, Tensor}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * Applies 1D max-pooling operation in kW regions by step size dW steps.
+ * Input sequence composed of nInputFrame frames.
+ * The input tensor in forward(input) is expected to be a 2D tensor (nInputFrame x inputFrameSize)
+ * or a 3D tensor (nBatchFrame x nInputFrame x inputFrameSize).
+ *
+ * If the input sequence is a 2D tensor of dimension nInputFrame x inputFrameSize,
+ * the output sequence will be nOutputFrame x inputFrameSize where
+ *
+ * nOutputFrame = (nInputFrame - kW) / dW + 1
+ *
+ * @param kW kernel width
+ * @param dW step size in width
+ * @tparam T The numeric type in the criterion, usually which are [[Float]] or [[Double]]
+ */
+class TemporalMaxPooling[T: ClassTag](
+  val kW: Int, var dW: Int = -1)
+  (implicit ev: TensorNumeric[T]) extends TensorModule[T] {
+  @transient var indices: Tensor[T] = _
+
+  if (dW == -1) {
+    dW = kW
+  }
+
+  def shapeCheck(
+    input: Tensor[T],
+    gradOutput: Tensor[T],
+    indices: Tensor[T],
+    kW: Int,
+    dW: Int
+  ): Unit = {
+    var dimS = 1 // sequence dimension
+    var dimF = 2 // feature dimension
+    val ndims = input.nDimension()
+
+    if (input.nDimension() == 3)
+    {
+      dimS = 2
+      dimF = 3
+    }
+
+    val niframe = input.size(dimS)
+    val framesize = input.size(dimF)
+    val noframe = (niframe - kW) / dW + 1
+
+    require(kW > 0,
+      s"kernel size should be greater than zero, but got kW: $kW")
+    require(dW > 0,
+      s"stride should be greater than zero, but got dW: $dW")
+
+    require(input.nDimension() == 2 || input.nDimension() == 3,
+      s"2D or 3D (batch mode) tensor expected for input, but got: ${input.nDimension()}")
+    require(input.size(dimS) >= kW,
+      s"input sequence smaller than kernel size. Got: ${input.size(dimS)}, Expected: $kW")
+
+    if (gradOutput != null) {
+      require(gradOutput.nDimension() == ndims && gradOutput.size(dimS) == noframe)
+      require(gradOutput.nDimension() == ndims && gradOutput.size(dimF) == framesize)
+    }
+    if (indices != null) {
+      require(indices.nDimension() == ndims && indices.size(dimS) == noframe)
+      require(indices.nDimension() == ndims && indices.size(dimF) == framesize)
+    }
+  }
+
+
+  def updateOutput(input: Tensor[T]): Tensor[T] = {
+    if (indices == null) {
+      indices = Tensor()
+    }
+
+    var nIFrame = 0
+    var nOFrame = 0
+    var frameSize = 0
+
+    var dimS = 1 // sequence dimension
+    var dimF = 2 // feature dimension
+
+    shapeCheck(input, null, null, kW, dW)
+
+    if (input.nDimension() == 3) {
+      dimS = 2
+      dimF = 3
+    }
+
+    nIFrame = input.size(dimS)
+    frameSize = input.size(dimF)
+    nOFrame = (nIFrame - kW) / dW + 1
+
+    val inputC = input.contiguous()
```

Comment:
use a buffer instead of contiguous(), since if it is not contiguous, it will create a new space each time